enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+(.spec.defaultNetwork.ovnKubernetesConfig) which will enable cluster-wide
+configuration of IPsec at cluster installation time:
+
+spec:
+  defaultNetwork:
+    type: OVNKubernetes
+    ovnKubernetesConfig:
+      mtu: 1400
+      genevePort: 6081
+      **enableIPsec: true/false**
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, I want to be able to enable IPsec at cluster installation
+time in the OVNKubernetes plugin in order to encrypt, authenticate and ensure
+the integrity of all pod-pod traffic.
+
+### Implementation Details/Notes/Constraints [optional]
+
+It is expected that the following changes will need to be made in order to
+enable this enhancement:
+
+- OVNKubernetes pod Dockerfiles will need to be updated in order to install
+  necessary components required by OVN/OVS to enable IPsec (e.g.
+  openvswitch-ipsec, libreswan, etc).
+- Modify Cluster Network Operator manifests to configure start up of packages
+  required by OVN/OVS in order to enable IPsec (e.g. ovs-monitor-ipsec, pluto,
+  etc) and determine how to containerize these components.
+- Ensure that all versions of all components are at the correct level required
+  by OVN/OVS. This may required back-porting of patches into RHEL. It has
+  already been identified that OVS and the Kernel will each require a patch to
+  enable this enhancement.
+- Ensure correct generation of the Certificate Authority (CA) certificate and
+  all node private keys and CA-signed certificates. It is envisioned that the
+  key generation mechanism (and keys) of OVN SSL will be reused.
+- Modify Cluster Network Operator manifests to ensure that all OVS/OVN
+  components are configured to use host private key, host certificate and CA
+  certificate.
+- Modify OVNKubernetes operator configuration object in order to allow
+  enablement of IPsec for the cluster.
+- Certificate rotation will be handled in the same as it is with the current
+  OVN TLS implementation.
+
+### Risks and Mitigations
+
+#### Upstreaming
+
+A number of changes will need to be made to upstream communities including the
+Linux kernel. As these may need time to execute, it is imperative that we
+identify them early and start the process.
+
+#### Performance
+
+Enablement of IPsec will cause degradation of overall networking throughput and
+latency due to the additional overhead of the IPsec protocol. Some early
+performance testing will be done in a non-production cluster in order to get
+an approximation of the expected performance degradation - These results will be
+posted at a later date. After implementation of the enhancement, it is expected
+that comprehensive benchmarking and profiling will take place. Based on the
+output of this activity, it may be decided to start some optimization activities
+in parallel to improve performance.
+
+#### Security Review
+
+See "Open Questions" below.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Will this need to go through some kind of Security Review?
```

Comment:
someone who actually knows what FIPS requires is going to have to look at it and say "yup, that's FIPS compliant" at some point...

Reply:
I'm not aware of what the OVN implementation of ipsec actually does. does it use pluto or does it implement the IKE protocol itself? If OVN is merely a configuration or orchestration layer for the tunnels, it might not need anything but configuration parameters to make sure that it's FIPS compliant.

One thing to note about FIPS compliance (and compliance in general), is that the traffic should be encrypted from installation and not as a day 2 operation; else the deployment is considered non-compliant.

Reply:
> I'm not aware of what the OVN implementation of ipsec actually does. does it use pluto or does it implement the IKE protocol itself? If OVN is merely a configuration or orchestration layer for the tunnels, it might not need anything but configuration parameters to make sure that it's FIPS compliant.

Looking at https://docs.openvswitch.org/en/latest/howto/ipsec/, it seems to be the former, so pluto handles the userspace keying making that part FIPS compliant in implementation.
How OVS handles interfacing and configuring pluto, might make a difference WRT deployment as you point out.

Reply:
Yes it merely orchestrates pluto. There is a separate daemon that monitors for changes in OVS configuration and then updates pluto by modifying its configuration files. If there are specific parameters that will ensure FIPS compliance, OVS could be updated to ensure they are used. @JAORMX could we discuss this further if you are knowledgeable in this area.

@danwinship I will specify this in this section.

Reply:
@markdgray sure. you can reach out to us in #forum-compliance in slack.